src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp

```diff
@@ -497,12 +504,12 @@ bool ConstantExpressionTemplate::parseLiteralAndAssertType(ReadBuffer & istr, co
 
     if (type_info.is_array || type_info.is_tuple || type_info.is_map)
     {
-        /// TODO faster way to check types without using Parsers
         ParserArrayOfLiterals parser_array;
         ParserTupleOfLiterals parser_tuple;
 
-        Tokens tokens_number(istr.position(), istr.buffer().end());
-        IParser::Pos iterator(tokens_number, static_cast<unsigned>(settings.max_parser_depth));
+        IParser::Pos iterator(token_iterator, static_cast<unsigned>(settings.max_parser_depth));
+        while (iterator->begin < istr.position())
+            ++iterator;
```

Comment:
Is there a way to do it faster or is it not significant?

Reply:
With this design I don't think it has any noticeable impact and it's the cost of caching the line tokens for later. On the other hand, an approach that **might** be faster would be to switch to always use the tokens for the rest of the line if you have already generated them, but that would complicate things even more.